src/kv/map.h

```diff
@@ -391,6 +327,125 @@ namespace kv
       return new Map(other, name, security_domain, replicated);
     }
 
+    void serialise(
+      const AbstractTxView* view,
+      KvStoreSerialiser& s,
+      bool include_reads) override
+    {
+      const auto committer =
+        dynamic_cast<const TxViewCommitter<K, V, H>*>(view);
+      if (committer == nullptr)
+      {
+        LOG_FAIL_FMT("Unable to serialise map due to type mismatch");
+        return;
+      }
+
+      const auto& change_set = committer->get_change_set();
+
+      s.start_map(name, security_domain);
+
+      if (include_reads)
+      {
+        s.serialise_read_version(change_set.read_version);
+
+        s.serialise_count_header(change_set.reads.size());
+        for (auto it = change_set.reads.begin(); it != change_set.reads.end();
+             ++it)
+        {
+          s.serialise_read(it->first, it->second);
+        }
+      }
+      else
+      {
+        s.serialise_read_version(NoVersion);
+        s.serialise_count_header(0);
+      }
+
+      uint64_t write_ctr = 0;
+      uint64_t remove_ctr = 0;
+      for (auto it = change_set.writes.begin(); it != change_set.writes.end();
+           ++it)
+      {
+        if (!is_deleted(it->second.version))
+        {
+          ++write_ctr;
+        }
+        else
+        {
+          auto search = roll->get_tail()->state.get(it->first);
+          if (search.has_value())
+          {
+            ++remove_ctr;
+          }
+        }
+      }
+
+      s.serialise_count_header(write_ctr);
+      for (auto it = change_set.writes.begin(); it != change_set.writes.end();
+           ++it)
+      {
+        if (!is_deleted(it->second.version))
+        {
+          s.serialise_write(it->first, it->second.value);
+        }
+      }
+
+      s.serialise_count_header(remove_ctr);
+      for (auto it = change_set.writes.begin(); it != change_set.writes.end();
+           ++it)
+      {
+        if (is_deleted(it->second.version))
+        {
+          s.serialise_remove(it->first);
+        }
+      }
+    }
+
+    AbstractTxView* deserialise(
+      KvStoreDeserialiser& d, Version version) override
+    {
+      // Create a new change set, and deserialise d's contents into it.
+      // TODO: Would really like this to be _only_ a TxViewCommitter, since we
```

Comment:
This would have been a neat separation of concerns (when deserialising, you build the `ChangeSet` and then the only thing you want to do with the resulting 'view' is commit it, so you shouldn't need a modifiable view?), but is not possible in the current code - PBFT's approach needs to access a real, deserialised-but-not-committed `TxView` from the same `Tx`. I've added some errors if this type-expectation invariant is broken in `tx.h`, and will remove this comment.